File: flipper/sanctum_rpg/pool.c
/*
 * pool.c — see pool.h. Spec 50 framework realization (slice 50.F0).
 *
 * Pure module. No Flipper SDK deps. Host-testable. Byte-equal across
 * hardware via integer-only FNV-1a chain transitions.
 */

#include "pool.h"

#include <stddef.h>
#include <string.h>

/* ─── Internal: FNV-1a 32-bit chain hash ──────────────────────────── */

static uint32_t fnv1a_mix(uint32_t h, uint32_t v) {
    h ^= (v & 0xFFu);        h *= 16777619u;
    h ^= ((v >> 8)  & 0xFFu); h *= 16777619u;
    h ^= ((v >> 16) & 0xFFu); h *= 16777619u;
    h ^= ((v >> 24) & 0xFFu); h *= 16777619u;
    return h;
}

static uint32_t fnv1a_seed(uint32_t a, uint32_t b, uint32_t c) {
    uint32_t h = 0x811C9DC5u;
    h = fnv1a_mix(h, a);
    h = fnv1a_mix(h, b);
    h = fnv1a_mix(h, c);
    /* avalanche */
    h ^= h >> 16;
    h *= 0x85EBCA6Bu;
    h ^= h >> 13;
    return h;
}

/* ─── Packed theme weights (16 × 4-bit values in 8 bytes) ──────────── */

uint8_t pool_theme_weight(const Pool* p, int theme_bit) {
    if(!p || theme_bit < 0 || theme_bit >= POOL_THEMES) return 0;
    uint8_t byte = p->theme_weights[theme_bit / 2];
    return (theme_bit & 1) ? (byte >> 4) : (byte & 0x0F);
}

static void pool_set_theme_weight(Pool* p, int theme_bit, uint8_t w) {
    if(!p || theme_bit < 0 || theme_bit >= POOL_THEMES) return;
    if(w > 15) w = 15;
    uint8_t* byte = &p->theme_weights[theme_bit / 2];
    if(theme_bit & 1) {
        *byte = (uint8_t)((*byte & 0x0F) | (w << 4));
    } else {
        *byte = (uint8_t)((*byte & 0xF0) | (w & 0x0F));
    }
}

/* ─── Convenience bias queries ─────────────────────────────────────── */

uint8_t pool_family_bias(const Pool* p, uint8_t family_id) {
    if(!p || family_id >= POOL_FAMILIES) return 0;
    return p->family_bias[family_id];
}

uint8_t pool_loot_kind_bias(const Pool* p, uint8_t kind_id) {
    if(!p || kind_id >= POOL_LOOT_KINDS) return 0;
    return p->loot_kind_bias[kind_id];
}

bool pool_equal(const Pool* a, const Pool* b) {
    if(!a || !b) return a == b;
    return memcmp(a, b, sizeof(Pool)) == 0;
}

/* ─── Biome primitive vocabulary (the user's locked sets) ─────────────
 * Cavern: ST=0, C=1, MC=2, LST=3, MW=4, reserved=5 (spec 50 §C.1).
 * Outdoor: SS=0, CA=1, MH=2, DF=3, SC=4, SR=5 (spec 50 §C.2).
 *
 * The vocabulary is UNIVERSAL — every save uses the same primitive IDs.
 * The per-save starting weights are hashed from prime_seed (below), so
 * each campaign's drift trajectory begins from a unique mix. */
#define BIOME_CAVERN  0u
#define BIOME_OUTDOOR 1u

#define CAVERN_VOCAB_COUNT  6
#define OUTDOOR_VOCAB_COUNT 6

static const uint8_t CAVERN_VOCAB[CAVERN_VOCAB_COUNT] = {0, 1, 2, 3, 4, 5};
static const uint8_t OUTDOOR_VOCAB[OUTDOOR_VOCAB_COUNT] = {0, 1, 2, 3, 4, 5};

static void pool_seed_bag(Pool* p, uint32_t prime_seed, uint8_t biome) {
    /* Initialize bag from biome vocabulary. The primitive IDs go in
     * declared order; the WEIGHTS are hashed from prime_seed so each
     * campaign starts uniquely. Slots beyond vocabulary length stay
     * empty (0xFF id). */
    const uint8_t* vocab = (biome == BIOME_OUTDOOR) ? OUTDOOR_VOCAB : CAVERN_VOCAB;
    uint8_t vlen = (biome == BIOME_OUTDOOR) ? OUTDOOR_VOCAB_COUNT
                                            : CAVERN_VOCAB_COUNT;
    for(int slot = 0; slot < POOL_BAG_SLOTS; slot++) {
        if(slot < vlen) {
            p->primitive_bag[slot][0] = vocab[slot];
            uint32_t h = fnv1a_seed(prime_seed, (uint32_t)biome,
                                    (uint32_t)(0xB00u | slot));
            /* Weight range 4..15 — every starting primitive has presence,
             * but the distribution is unique per save. */
            p->primitive_bag[slot][1] = (uint8_t)(4 + (h % 12u));
        } else {
            p->primitive_bag[slot][0] = 0xFF;
            p->primitive_bag[slot][1] = 0;
        }
    }
}

/* ─── pool_at_prime: initial state at the prime stamp ──────────────── */

void pool_at_prime(uint32_t prime_seed, uint8_t biome, Pool* out) {
    if(!out) return;
    memset(out, 0, sizeof(*out));

    /* Theme weights — 16 themes, each hashed to a 0..15 value. Every save's
     * theme distribution at the prime is unique. */
    for(int t = 0; t < POOL_THEMES; t++) {
        uint32_t h = fnv1a_seed(prime_seed, (uint32_t)biome,
                                (uint32_t)(0xA00u | t));
        pool_set_theme_weight(out, t, (uint8_t)(h % 16u));
    }

    /* Primitive bag — biome vocabulary with hashed starting weights. */
    pool_seed_bag(out, prime_seed, biome);

    /* Family bias — 5 families, hashed to 0..7 each. Sum is unbounded
     * (each is additive into the family_roll weight walk). */
    for(int f = 0; f < POOL_FAMILIES; f++) {
        uint32_t h = fnv1a_seed(prime_seed, (uint32_t)biome,
                                (uint32_t)(0xC00u | f));
        out->family_bias[f] = (uint8_t)(h % 8u);
    }

    /* Loot kind bias — 7 kinds, hashed to 0..5 each. */
    for(int k = 0; k < POOL_LOOT_KINDS; k++) {
        uint32_t h = fnv1a_seed(prime_seed, (uint32_t)biome,
                                (uint32_t)(0xD00u | k));
        out->loot_kind_bias[k] = (uint8_t)(h % 6u);
    }

    out->intensity = 0;
    out->rotation_bias = 0;
    out->pedigree = fnv1a_seed(prime_seed, (uint32_t)biome, 0u);
    out->last_drop_slot = 0xFFu; /* spec 50 addendum §F1.A.4 — no prior step */
}

/* ─── pool_step: advance one chunk in a direction ─────────────────── */

void pool_step(Pool* p, uint32_t prime_seed, int dir) {
    if(!p) return;
    if(dir < 0 || dir > 3) return;

    /* Advance pedigree first — every subsequent drift reads from it. */
    p->pedigree = fnv1a_mix(
        fnv1a_mix(p->pedigree, (uint32_t)dir),
        prime_seed);

    /* Intensity grows with every step (saturating cap per spec 50 §9.4). */
    if(p->intensity < 255) p->intensity++;

    /* Rotation bias drifts by direction (mod 4). */
    p->rotation_bias = (uint8_t)((p->rotation_bias + (uint8_t)dir + 1u) & 3u);

    /* Drift theme weights: small +1/-1 nudges driven by pedigree bits.
     * Each step touches up to 4 theme slots so drift is gradual. */
    for(int n = 0; n < 4; n++) {
        uint32_t h = fnv1a_mix(p->pedigree, (uint32_t)(0xE00u | n));
        int slot = (int)(h & 0x0Fu);
        int delta = ((h >> 8) & 1u) ? +1 : -1;
        int cur = (int)pool_theme_weight(p, slot);
        cur += delta;
        if(cur < 0) cur = 0;
        if(cur > 15) cur = 15;
        pool_set_theme_weight(p, slot, (uint8_t)cur);
    }

    /* Drift family bias: pick one slot, ±1. */
    {
        uint32_t h = fnv1a_mix(p->pedigree, 0xF1F1u);
        int slot = (int)(h % POOL_FAMILIES);
        int delta = ((h >> 16) & 1u) ? +1 : -1;
        int cur = (int)p->family_bias[slot] + delta;
        if(cur < 0) cur = 0;
        if(cur > 15) cur = 15;
        p->family_bias[slot] = (uint8_t)cur;
    }

    /* Drift loot kind bias: pick one slot, ±1. */
    {
        uint32_t h = fnv1a_mix(p->pedigree, 0xF2F2u);
        int slot = (int)(h % POOL_LOOT_KINDS);
        int delta = ((h >> 16) & 1u) ? +1 : -1;
        int cur = (int)p->loot_kind_bias[slot] + delta;
        if(cur < 0) cur = 0;
        if(cur > 15) cur = 15;
        p->loot_kind_bias[slot] = (uint8_t)cur;
    }

    /* Rotate primitive bag — the "recycled to be re-assigned" mechanic
     * from the user's design (spec 50 §0). Per spec 50 addendum §F1.A:
     *
     *   DROP: pick the lowest-weight occupied slot, EXCLUDING (a) slots
     *   at BAG_WEIGHT_FLOOR (every primitive keeps presence) and (b) the
     *   slot drained on the previous step (anti-recurrence). Tiebreaker
     *   is a hashed rotor over the pedigree, NOT lowest-index — index-
     *   tiebreaker caused a drain-cascade where one slot bled to zero
     *   forever within ~30 steps.
     *
     *   BUMP: pick a hashed target slot; if it collides with the drop
     *   slot, walk forward through the bag until a different occupied
     *   slot is found.
     *
     * The bag's CONTENT (primitive IDs) stays universal; only WEIGHTS drift. */
    {
        /* --- DROP --- */
        int drop = -1;
        uint8_t low_w = 255;
        uint8_t rotor = (uint8_t)((p->pedigree >> 24) & 7u);
        for(int r = 0; r < POOL_BAG_SLOTS; r++) {
            int s = (int)((rotor + (uint8_t)r) & 7u);
            uint8_t id = p->primitive_bag[s][0];
            uint8_t w  = p->primitive_bag[s][1];
            if(id == 0xFF) continue;
            if(w <= BAG_WEIGHT_FLOOR) continue;
            if((uint8_t)s == p->last_drop_slot) continue;
            if(w < low_w) { low_w = w; drop = s; }
        }
        if(drop >= 0) {
            p->primitive_bag[drop][1]--;
            p->last_drop_slot = (uint8_t)drop;
        }
        /* else: no eligible candidate — bag is floor-locked this step. */

        /* --- BUMP --- */
        uint32_t h_bump = fnv1a_mix(p->pedigree, 0xF3F3u);
        int bump = (int)(h_bump % POOL_BAG_SLOTS);
        for(int r = 0; r < POOL_BAG_SLOTS; r++) {
            int s = (int)(((uint32_t)bump + (uint32_t)r) % POOL_BAG_SLOTS);
            if(p->primitive_bag[s][0] == 0xFF) continue;
            if(s == drop) continue;
            bump = s;
            break;
        }
        if(bump >= 0 && p->primitive_bag[bump][1] < 15) {
            p->primitive_bag[bump][1]++;
        }
    }
}

/* ─── pool_at: canonical X-first-then-Y walk from prime ─────────────── */

void pool_at(uint32_t prime_seed, uint8_t biome, int chunk_x, int chunk_y,
             Pool* out) {
    if(!out) return;
    pool_at_prime(prime_seed, biome, out);

    /* X first — east (+x) or west (-x). */
    int x_steps = (chunk_x >= 0) ? chunk_x : -chunk_x;
    int x_dir = (chunk_x >= 0) ? POOL_DIR_EAST : POOL_DIR_WEST;
    for(int i = 0; i < x_steps; i++) {
        pool_step(out, prime_seed, x_dir);
    }

    /* Y second — south (+y) or north (-y). */
    int y_steps = (chunk_y >= 0) ? chunk_y : -chunk_y;
    int y_dir = (chunk_y >= 0) ? POOL_DIR_SOUTH : POOL_DIR_NORTH;
    for(int i = 0; i < y_steps; i++) {
        pool_step(out, prime_seed, y_dir);
    }
}

File: core/systems/determinism/pool.py
"""
pool.py — Python conformance kernel for flipper/sanctum_rpg/pool.c

Reproduces pool_at_prime / pool_at / pool_step field-for-field.
Integer-only. Zero floats. Named constants at file top.

Pool is represented as a plain dataclass matching the JSON schema.
Do NOT raw-memcpy Pool across languages — use field-wise accessors.
sizeof(Pool) == 48 in C (not 42 as the header comment claims; 3 bytes padding
after rotation_bias before pedigree, 3 bytes after last_drop_slot).

Spec: sanctum-os/docs/specs/53_seed_is_truth.md §9-A
Conformance: tests/test_conformance_rng_pool.py
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List

# ── Constants transcribed verbatim from pool.h and pool.c ────────────────────
POOL_THEMES: int = 16         # theme vocabulary size
POOL_BAG_SLOTS: int = 8       # primitive bag capacity
POOL_FAMILIES: int = 5        # creature families per biome
POOL_LOOT_KINDS: int = 7      # loot kind count
BAG_WEIGHT_FLOOR: int = 1     # minimum bag slot weight (anti-drain)
BAG_WEIGHT_MAX: int = 15      # maximum bag slot weight (cap for bump)

# Walk direction encoding (pool.h)
POOL_DIR_EAST: int = 0
POOL_DIR_WEST: int = 1
POOL_DIR_SOUTH: int = 2
POOL_DIR_NORTH: int = 3

# FNV-1a 32-bit parameters
_FNV_OFFSET_BASIS: int = 0x811C9DC5
_FNV_PRIME: int = 16777619        # 0x01000193

# FNV-1a avalanche multiplier used in fnv1a_seed (pool.c line ~30).
# LOOKALIKE HAZARD: this is 0x85EBCA6B.
# rng.c chunk-coord multiplier uses 0x85EBCA77 (different).
# rng.c fmix second constant uses 0x846CA68B (different).
_FNV_AVALANCHE_MULT: int = 0x85EBCA6B

# Biome IDs matching pool.c
_BIOME_CAVERN: int = 0
_BIOME_OUTDOOR: int = 1

# Biome primitive vocabularies (pool.c CAVERN_VOCAB / OUTDOOR_VOCAB)
_CAVERN_VOCAB: tuple[int, ...] = (0, 1, 2, 3, 4, 5)
_OUTDOOR_VOCAB: tuple[int, ...] = (0, 1, 2, 3, 4, 5)

# u32 mask
_U32_MASK: int = 0xFFFFFFFF


# ── u32 helper ───────────────────────────────────────────────────────────────

def _u32(x: int) -> int:
    """Truncate to uint32_t. Apply after every multiply."""
    return x & _U32_MASK


# ── FNV-1a internals ─────────────────────────────────────────────────────────

def _fnv1a_mix(h: int, v: int) -> int:
    """
    FNV-1a 32-bit chain mix — feeds each byte of v low-byte-first.
    Transcribed from pool.c fnv1a_mix().
    """
    h ^= (v & 0xFF);         h = _u32(h * _FNV_PRIME)
    h ^= ((v >> 8)  & 0xFF); h = _u32(h * _FNV_PRIME)
    h ^= ((v >> 16) & 0xFF); h = _u32(h * _FNV_PRIME)
    h ^= ((v >> 24) & 0xFF); h = _u32(h * _FNV_PRIME)
    return h


def _fnv1a_seed(a: int, b: int, c: int) -> int:
    """
    FNV-1a 32-bit seed from three uint32 inputs.
    Starts from offset basis, mixes a/b/c in order, then avalanche pass.
    Transcribed from pool.c fnv1a_seed().
    """
    h = _FNV_OFFSET_BASIS
    h = _fnv1a_mix(h, a)
    h = _fnv1a_mix(h, b)
    h = _fnv1a_mix(h, c)
    # avalanche: h ^= h>>16; h *= 0x85EBCA6B; h ^= h>>13
    h ^= (h >> 16)
    h = _u32(h * _FNV_AVALANCHE_MULT)
    h ^= (h >> 13)
    return h


# ── Pool dataclass ────────────────────────────────────────────────────────────

@dataclass
class Pool:
    """
    Python representation of the C Pool struct — field-wise, not packed bytes.
    All fields match the JSON schema in golden_vectors.json.
    """
    # 16 theme weights, each 0..15 (unpacked from C's nibble-packed 8-byte form)
    theme_weights: List[int] = field(default_factory=lambda: [0] * POOL_THEMES)
    # primitive_bag[slot] = [id, weight]; id 0xFF = empty
    primitive_bag: List[List[int]] = field(
        default_factory=lambda: [[0xFF, 0] for _ in range(POOL_BAG_SLOTS)]
    )
    family_bias: List[int] = field(default_factory=lambda: [0] * POOL_FAMILIES)
    loot_kind_bias: List[int] = field(default_factory=lambda: [0] * POOL_LOOT_KINDS)
    intensity: int = 0
    rotation_bias: int = 0
    pedigree: int = 0
    last_drop_slot: int = 0xFF


def pool_theme_weight(p: Pool, theme_bit: int) -> int:
    """Read theme weight at index. Returns 0 on out-of-range (mirrors C behaviour)."""
    if theme_bit < 0 or theme_bit >= POOL_THEMES:
        return 0
    return p.theme_weights[theme_bit]


def _pool_set_theme_weight(p: Pool, theme_bit: int, w: int) -> None:
    """Write theme weight, clamped to [0..15]."""
    if theme_bit < 0 or theme_bit >= POOL_THEMES:
        return
    p.theme_weights[theme_bit] = max(0, min(15, w))


# ── pool_at_prime ─────────────────────────────────────────────────────────────

def pool_at_prime(prime_seed: int, biome: int) -> Pool:
    """
    Initialize Pool at the campaign prime stamp.
    Same (prime_seed, biome) → identical Pool.
    Transcribed from pool.c pool_at_prime().
    """
    p = Pool()

    # Theme weights: 16 themes hashed to 0..15 each
    for t in range(POOL_THEMES):
        h = _fnv1a_seed(prime_seed, biome, 0xA00 | t)
        _pool_set_theme_weight(p, t, h % 16)

    # Primitive bag: biome vocabulary with hashed starting weights
    vocab = _OUTDOOR_VOCAB if biome == _BIOME_OUTDOOR else _CAVERN_VOCAB
    for slot in range(POOL_BAG_SLOTS):
        if slot < len(vocab):
            p.primitive_bag[slot][0] = vocab[slot]
            h = _fnv1a_seed(prime_seed, biome, 0xB00 | slot)
            # Weight range 4..15
            p.primitive_bag[slot][1] = 4 + (h % 12)
        else:
            p.primitive_bag[slot][0] = 0xFF
            p.primitive_bag[slot][1] = 0

    # Family bias: 5 families hashed to 0..7 each
    for f in range(POOL_FAMILIES):
        h = _fnv1a_seed(prime_seed, biome, 0xC00 | f)
        p.family_bias[f] = h % 8

    # Loot kind bias: 7 kinds hashed to 0..5 each
    for k in range(POOL_LOOT_KINDS):
        h = _fnv1a_seed(prime_seed, biome, 0xD00 | k)
        p.loot_kind_bias[k] = h % 6

    p.intensity = 0
    p.rotation_bias = 0
    p.pedigree = _fnv1a_seed(prime_seed, biome, 0)
    p.last_drop_slot = 0xFF  # spec 50 addendum §F1.A.4: no prior step

    return p


# ── pool_step ────────────────────────────────────────────────────────────────

def pool_step(p: Pool, prime_seed: int, direction: int) -> None:
    """
    Advance Pool one step in direction (POOL_DIR_*). Mutates p in place.

    Execution order MUST match pool.c exactly:
      1. pedigree advance
      2. intensity saturate
      3. rotation_bias
      4. theme drift (4 rounds, reads live array each round)
      5. family drift
      6. loot drift
      7. bag drop + bump
    """
    if direction < 0 or direction > 3:
        return

    # Step 1: pedigree — inner is dir, outer is prime_seed
    p.pedigree = _fnv1a_mix(_fnv1a_mix(p.pedigree, direction), prime_seed)

    # Step 2: intensity — saturating uint8
    if p.intensity < 255:
        p.intensity += 1

    # Step 3: rotation_bias
    p.rotation_bias = (p.rotation_bias + direction + 1) & 3

    # Step 4: theme drift — 4 rounds, each reads the LIVE array (writes visible to next round)
    for n in range(4):
        h = _fnv1a_mix(p.pedigree, 0xE00 | n)
        slot = h & 0x0F
        delta = +1 if ((h >> 8) & 1) else -1
        cur = pool_theme_weight(p, slot) + delta
        _pool_set_theme_weight(p, slot, cur)

    # Step 5: family drift — one slot ±1, clamp [0..15]
    h = _fnv1a_mix(p.pedigree, 0xF1F1)
    fam_slot = h % POOL_FAMILIES
    fam_delta = +1 if ((h >> 16) & 1) else -1
    p.family_bias[fam_slot] = max(0, min(15, p.family_bias[fam_slot] + fam_delta))

    # Step 6: loot drift — one slot ±1, clamp [0..15]
    h = _fnv1a_mix(p.pedigree, 0xF2F2)
    loot_slot = h % POOL_LOOT_KINDS
    loot_delta = +1 if ((h >> 16) & 1) else -1
    p.loot_kind_bias[loot_slot] = max(0, min(15, p.loot_kind_bias[loot_slot] + loot_delta))

    # Step 7: bag rotation — DROP then BUMP
    # --- DROP ---
    # Rotor-order scan: iterate slots in order (rotor + r) & 7.
    # Strict less-than tiebreak: earliest-in-rotor-order wins among ties.
    # Skip: id == 0xFF (empty), weight <= BAG_WEIGHT_FLOOR, slot == last_drop_slot.
    drop = -1
    low_w = 256  # sentinel above max uint8
    rotor = (p.pedigree >> 24) & 7
    for r in range(POOL_BAG_SLOTS):
        s = (rotor + r) & 7
        bag_id = p.primitive_bag[s][0]
        bag_w = p.primitive_bag[s][1]
        if bag_id == 0xFF:
            continue
        if bag_w <= BAG_WEIGHT_FLOOR:
            continue
        if s == p.last_drop_slot:
            continue
        if bag_w < low_w:  # STRICT less-than: first tie in rotor order wins
            low_w = bag_w
            drop = s

    if drop >= 0:
        p.primitive_bag[drop][1] -= 1
        p.last_drop_slot = drop
    # else: no eligible candidate — bag floor-locked this step

    # --- BUMP ---
    # h_bump selects starting slot; walk forward until occupied and != drop.
    h_bump = _fnv1a_mix(p.pedigree, 0xF3F3)
    bump = h_bump % POOL_BAG_SLOTS
    for r in range(POOL_BAG_SLOTS):
        s = (bump + r) % POOL_BAG_SLOTS
        if p.primitive_bag[s][0] == 0xFF:
            continue
        if s == drop:
            continue
        bump = s
        break

    if bump >= 0 and p.primitive_bag[bump][1] < BAG_WEIGHT_MAX:
        p.primitive_bag[bump][1] += 1


# ── pool_at ───────────────────────────────────────────────────────────────────

def pool_at(prime_seed: int, biome: int, chunk_x: int, chunk_y: int) -> Pool:
    """
    Compute Pool at chunk (chunk_x, chunk_y) via canonical X-first-then-Y walk.
    Same arguments → identical Pool. Transcribed from pool.c pool_at().
    """
    p = pool_at_prime(prime_seed, biome)

    # X first — east (+x) or west (-x)
    x_steps = chunk_x if chunk_x >= 0 else -chunk_x
    x_dir = POOL_DIR_EAST if chunk_x >= 0 else POOL_DIR_WEST
    for _ in range(x_steps):
        pool_step(p, prime_seed, x_dir)

    # Y second — south (+y) or north (-y)
    y_steps = chunk_y if chunk_y >= 0 else -chunk_y
    y_dir = POOL_DIR_SOUTH if chunk_y >= 0 else POOL_DIR_NORTH
    for _ in range(y_steps):
        pool_step(p, prime_seed, y_dir)

    return p
